src/config.ts

```diff
@@ -118,19 +118,41 @@ export class CacheConfig {
     let lockHash = core.getState(STATE_LOCKFILE_HASH);
     let keyFiles: Array<string> = JSON.parse(core.getState(STATE_LOCKFILES) || "[]");
 
-    if (!lockHash) {
-      const globber = await glob.create("**/Cargo.toml\n**/Cargo.lock\nrust-toolchain\nrust-toolchain.toml", {
-        followSymbolicLinks: false,
-      });
-      keyFiles = await globber.glob();
-      keyFiles.sort((a, b) => a.localeCompare(b));
+    // Constructs the workspace config and paths to restore:
+    // The workspaces are given using a `$workspace -> $target` syntax.
 
+    const workspaces: Array<Workspace> = [];
+    const workspacesInput = core.getInput("workspaces") || ".";
+    for (const workspace of workspacesInput.trim().split("\n")) {
+      let [root, target = "target"] = workspace.split("->").map((s) => s.trim());
+      root = path.resolve(root);
+      target = path.join(root, target);
+      workspaces.push(new Workspace(root, target));
+    }
+    self.workspaces = workspaces;
+
+    if (!lockHash) {
       hasher = crypto.createHash("sha1");
-      for (const file of keyFiles) {
-        for await (const chunk of fs.createReadStream(file)) {
-          hasher.update(chunk);
+
+      async function globHash(pattern: string): Promise<void> {
+        const globber = await glob.create(pattern, {
+          followSymbolicLinks: false,
+        });
+        keyFiles = keyFiles.concat(await globber.glob());
+        keyFiles.sort((a, b) => a.localeCompare(b));
+
+        for (const file of keyFiles) {
+          for await (const chunk of fs.createReadStream(file)) {
+            hasher.update(chunk);
+          }
         }
       }
+
+      await globHash("rust-toolchain\nrust-toolchain.toml");
```

Comment:
I’m wondering, is the `rust-toolchain` file something workspace specific? I think so?

Reply:
You're right, cargo uses the closest one but it'll search all the way to the root.